pkg/forwarders/rawchannelforwarder.go

```diff
@@ -0,0 +1,25 @@
+package forwarders
+
+import (
+	"github.com/kong/kubernetes-telemetry/pkg/types"
+)
+
+type rawChannelForwarder struct {
+	ch chan types.Report
+}
+
+// NewRawChannelForwarder creates new rawChannelForwarder.
+func NewRawChannelForwarder(ch chan types.Report) *rawChannelForwarder {
+	return &rawChannelForwarder{
+		ch: ch,
+	}
+}
+
+func (f *rawChannelForwarder) Name() string {
+	return "LogForwarder"
+}
+
+func (f *rawChannelForwarder) Forward(r types.Report) error {
+	f.ch <- r
+	return nil
+}
```

Comment:
So... yes. When the caller (owner of the `ch` channel) is not taking items of the channel in a timely manner then this could cause problems.

Using a buffered channel in my view would only delay the problem; well only if the problem is permanent or lasts longer than filling of said buffer. For spurious blockages this would technically help.

The problem with this type of usage of `select` (which I guess you're trying to advocate for) is that `select` in general is not deterministic so we can't really use it as:

```go
select {
	case f.ch <- r:
	default:
}
```

because we'd randomly get either of the 2 branches if the `f.ch` was in fact able to receive data.

Changing `default` to a "channel with a timeout" construct feels a bit wrong. Perhaps changing the interface of the forwarder altogether, to include a `context.Context` param so that we can use `ctx.Done()` in a `select` would make at least a little improvement in this aspect.

I'm open to ideas.

Reply:
> Using a buffered channel in my view would only delay the problem; well only if the problem is permanent or lasts longer than filling of said buffer. For spurious blockages this would technically help.

It would also make it possible to `select` on the channel, meaning that it doesn't only delay the problem, it allows calling code to identify the problem.

> The problem with this type of usage of select (which I guess you're trying to advocate for) is that select in general is not deterministic so we can't really use it because we'd randomly get either of the 2 branches if the f.ch was in fact able to receive data.

I'm a bit confused by this comment. What I think you're referring to is the pseudo-random selection of case statements when two or more case statements can proceed simultaneously, as defined [here in the spec](https://go.dev/ref/spec#Select_statements)? To my knowledge (and if I'm understanding the refresher I just did right) this only refers to the case statements in a select after the check to see if any case statements can proceed fails, so if no case can proceed during processing, it falls through to the default case? So the example should in fact work? :thinking: 

> Changing default to a "channel with a timeout" construct feels a bit wrong. Perhaps changing the interface of the forwarder altogether, to include a context.Context param so that we can use ctx.Done() in a select would make at least a little improvement in this aspect.

When I originally threw out the idea, this was actually what I intended just with the specific example of a context with a timeout in mind.

> I'm open to ideas.

End of the day my main concern is trying to get ahead of any problems we haven't fathomed yet, as this tool should ultimately be innocuous to load. In my mind removing any and all ambiguity about how public functions and methods should operate is a decent way to avoid some issues, if nothing else so that they can bubble up problems quicker. On the other hand I don't see any obvious places where this machinery would start to jam up so for now I consider this comment resolvable at your discretion, and would suggest that we add some additional testing that tries to instigate a "jammed up" pipeline for reports to see if that exercise highlights any rough edges for us.

Reply:
As for my example, let's give this full code snippet a closer look.

```go
package main

import "fmt"

func main() {
	var (
		cases int
		defaults int
	)

	ch := make(chan struct{})
	close(ch)

	for i := 0; i < 1_000_000; i++ {
		select {
		case <-ch:
			cases++
		default:
			defaults++
		}
	}

	fmt.Printf("cases: %d\n", cases)
	fmt.Printf("defaults: %d\n", defaults)
}
```

This will always print

```
cases: 1000000
defaults: 0
```

(which I was actually surprised about: I remembered that in such cases `default` can also fire). Which makes sense I guess reading the spec:

> 2. If one or more of the communications can proceed, a single one that can proceed is chosen via a uniform pseudo-random selection. Otherwise, if there is a default case, that case is chosen. If there is no default case, the "select" statement blocks until at least one of the communications can proceed.

So only If we'd added another channel to that select statement (could be another closed channel) we'd get random choice between the 2.

On the other hand, if you'd run the following snippet:

```go
package main

import "fmt"

func main() {
	var (
		cases    int
		defaults int
	)

	ch := make(chan struct{})
	go func() {
		for {
			ch <- struct{}{}
		}
	}()

	for i := 0; i < 1_000_000; i++ {
		select {
		case <-ch:
			cases++
		default:
			defaults++
		}
	}

	fmt.Printf("cases: %d\n", cases)
	fmt.Printf("defaults: %d\n", defaults)
}
```

You'd get results like these:

```
cases: 931
defaults: 999069
```

which make sense because the "sender loop" does not keep up with sending "as fast as the default case can be selected". And this is the one that I was "afraid about": i.e. having the consumer on the other end "miss" an entry because it didn't manage to be ready to receive instantaneously.

> > I'm open to ideas.
> 
> End of the day my main concern is trying to get ahead of any problems we haven't fathomed yet, as this tool should ultimately be innocuous to load. In my mind removing any and all ambiguity about how public functions and methods should operate is a decent way to avoid some issues, if nothing else so that they can bubble up problems quicker. On the other hand I don't yet see any obvious places where this machinery would start to jam up and I don't want to hang this PR up over something without concrete examples handy so for now I consider this comment resolvable at your discretion, and would suggest that we add some additional testing that tries to instigate a "jammed up" pipeline for reports to see if that exercise highlights any rough edges for us.

I'll take note of that and I will try to break it if I find a moment and an idea on how can I achieve this. :)

Reply:
I've created an issue to address this: https://github.com/Kong/kubernetes-telemetry/issues/47.

Let's close this discussion (as you suggested) and address it there.

Reply:
Your examples make perfect sense to me, and now I think I see what you were getting at, particularly with this comment:

> And this is the one that I was "afraid about": i.e. having the consumer on the other end "miss" an entry because it didn't manage to be ready to receive instantaneously.

So it wasn't that the default case is part of the pseudo-random selection of cases when actual case statements are valid (because that doesn't happen, as documented and as you provided an example for) you're worried that the default case would be hit in the interim while the channels are busy. That would indeed happen and particularly because you used an unbuffered channel and is part of why I was suggesting using/requiring a buffered channel AND clear error messages letting the caller know the channel is full. This however was just one suggestion of how to alleviate the potential problems, and I suppose neither here nor there for the moment.

As you suggested, I'm OK with using what we have for now and seeing how it behaves and following up to check on it in #47.